pkg/products/rhssouser/reconciler.go

```diff
@@ -243,9 +268,342 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, installation *inte
+		return integreatlyv1alpha1.PhaseFailed, errors.Wrap(err, "failed to list the keycloak users")
+	}
+
+	// Sync keycloak with openshift users
+	users, err := syncAdminUsersInMasterRealm(keycloakUsers, ctx, serverClient, r.Config.GetNamespace())
+	if err != nil {
+		return integreatlyv1alpha1.PhaseFailed, errors.Wrap(err, "failed to synchronize the users")
+	}
+
+	// Create / update the synchronized users
+	for _, user := range users {
+		or, err = r.createOrUpdateKeycloakAdmin(user, installation, ctx, serverClient)
+		if err != nil {
+			return integreatlyv1alpha1.PhaseFailed, errors.Wrap(err, "failed to create/update the customer admin user")
+		} else {
+			r.logger.Infof("The operation result for keycloakuser %s was %s", user.UserName, or)
+		}
+	}
+
 	return integreatlyv1alpha1.PhaseCompleted, nil
 }
 
+func identityProviderExists(kcClient keycloakCommon.KeycloakInterface) (bool, error) {
+	provider, err := kcClient.GetIdentityProvider(idpAlias, "master")
+	if err != nil {
+		return false, err
+	}
+	if provider != nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+// The master realm will be created as part of the Keycloak install. Here we update it to add the openshift idp
+func (r *Reconciler) updateMasterRealm(ctx context.Context, serverClient k8sclient.Client, installation *integreatlyv1alpha1.Installation, host string) (*keycloak.KeycloakRealm, error) {
+	keycloakMasterRealmName := "master"
+
+	kcr := &keycloak.KeycloakRealm{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      keycloakMasterRealmName,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+
+	or, err := controllerutil.CreateOrUpdate(ctx, serverClient, kcr, func() error {
+		kcr.Spec.RealmOverrides = []*keycloak.RedirectorIdentityProviderOverride{
+			{
+				IdentityProvider: idpAlias,
+				ForFlow:          "browser",
+			},
+		}
+
+		kcr.Spec.InstanceSelector = &metav1.LabelSelector{
+			MatchLabels: getInstanceLabels(),
+		}
+
+		kcr.Labels = getMasterLabels()
+
+		kcr.Spec.Realm = &keycloak.KeycloakAPIRealm{
+			ID:          keycloakMasterRealmName,
+			Realm:       keycloakMasterRealmName,
+			Enabled:     true,
+			DisplayName: keycloakMasterRealmName,
+		}
+
+		// The identity providers need to be set up before the realm CR gets
+		// created because the Keycloak operator does not allow updates to
+		// the realms
+		err := r.setupOpenshiftIDP(ctx, installation, kcr, serverClient, "master", host)
+		if err != nil {
+			return errors.Wrap(err, "failed to setup Openshift IDP for user-sso")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create/update keycloak realm: %w", err)
+	}
+	r.logger.Infof("The operation result for keycloakrealm %s was %s", kcr.Name, or)
+
+	return kcr, nil
+}
+
+func (r *Reconciler) createOrUpdateKeycloakAdmin(user keycloak.KeycloakAPIUser, inst *integreatlyv1alpha1.Installation, ctx context.Context, serverClient k8sclient.Client) (controllerutil.OperationResult, error) {
+	selector := &keycloak.KeycloakUser{
```

Comment:
`selector` is a bit weird name for KeycloakUser, code inside of controllerutil.CreateOrUpdate was a bit confusing because of that.

Reply:
updated to kcUser